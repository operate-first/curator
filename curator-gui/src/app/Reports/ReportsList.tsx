import React from 'react';
import '@patternfly/react-core/dist/styles/base.css';
import { Table, TableHeader, TableBody, TableVariant, BodyCell } from '@patternfly/react-table';
import { Link } from 'react-router-dom';
import { dataObject } from './ReportsDataFilterForm';

type myProps = {
  columnTitle: object;
  tableData: Array<dataObject>;
  startDate: Date;
};

type myState = {
  columns: Array<object | string>;
  rows: Array<row>;
};

type row = {
  cells: cells;
};

type cells = Array<JSX.Element | number | string>;

/*
The table/list which displays metering report data for the Standard Report Type.
Currently displays columns for namespace, cpu usage, network usage, and memory usage.
*/
class ReportsList extends React.Component<myProps, myState> {
  constructor(myProps) {
    super(myProps);

    const rowData: Array<row> = [];
    myProps.tableData.forEach(dataRow => {
      rowData.push({
        cells: [
          <BodyCell key={`/projectlist/${dataRow['namespace']}`}>
            <Link to={`/projectlist/${dataRow['namespace']}`} key={`/projectlist/${dataRow['namespace']}`}>
              {dataRow['namespace']}
            </Link>
          </BodyCell>,
          dataRow['podUsageCpuCoreSeconds'],

        ]
      });
    })

    this.state = {
      columns: [
        myProps.columnTitle['namespace'],
        myProps.columnTitle['podUsageCpuCoreSeconds']
      ],
      rows: rowData
    };
  }

  shouldComponentUpdate(nextProps: myProps,nextState: myState){
    console.log(nextProps.tableData);
    console.log(JSON.stringify(nextProps)!== JSON.stringify(this.props))
    return JSON.stringify(nextProps)!== JSON.stringify(this.props)
  }

  UNSAFE_componentWillReceiveProps(nextProps: myProps) {
      const rowData: Array<row> = [];

      nextProps.tableData.forEach(dataRow => {
        rowData.push({
          cells: [
            <BodyCell key={`/projectlist/${dataRow['namespace']}`}>
              <Link to={`/projectlist/${dataRow['namespace']}`} key={`/projectlist/${dataRow['namespace']}`}>
              {dataRow['namespace']}
              </Link>
            </BodyCell>,
            dataRow['podUsageCpuCoreSeconds']
          ]
        });
      });
      this.setState({ ...this.state, rows: rowData });
  }

  render() {
    return (
      <div>
        <Table
          key={'dataTable'}
          aria-label="Compact Table"
          variant={TableVariant.compact}
          cells={this.state.columns}
          rows={this.state.rows}
          caption="List of reports"
        >
          <TableHeader />
          <TableBody />
        </Table>
      </div>
    );
  }
}
export default ReportsList;
