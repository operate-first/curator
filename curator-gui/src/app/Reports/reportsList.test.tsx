import React from 'react';
import { shallow } from 'enzyme';
import ReportsList from './ReportsList';

describe('ReportsList tests', () => {
   
  // ReportsList tests. These tests check against a ReportsList component snapshot 
  // that is generated with sample test data. 
  test('should render default DashboardTable component', () => {
    const testDate = new Date();
    const testTableData = 
    [{namespace: "openshift-monitoring", podUsageCpuCoreSeconds: 21.1, network: 12, memory: 0},
    {namespace: "hello-world", podUsageCpuCoreSeconds: 13.4, network: 5, memory: 0},
    {namespace: "myproject", podUsageCpuCoreSeconds: 22.4, network: 0, memory: 0},
    {namespace: "sample-project", podUsageCpuCoreSeconds: 33.5, network: 6, memory: 0},
    {namespace: "sample-project-1", podUsageCpuCoreSeconds: 16.1, network: 7, memory: 0},
    {namespace: "sample-project-2", podUsageCpuCoreSeconds: 17.4, network: 8, memory: 0},
    {namespace: "sample-project-3", podUsageCpuCoreSeconds: 15.5, network: 9, memory: 0}]
    const testColumnTitle = {
        namespace: 'Report Name',
        podUsageCpuCoreSeconds: 'Pod CPU Usage in Seconds',
        network: 'Network Usage in Megabits per Second',
        memory: 'Memory Usage in Gigabytes'
    }
    const view = shallow(<ReportsList key={'ReportsList'} startDate={testDate}
        tableData={testTableData} columnTitle={testColumnTitle}/>);
    expect(view).toMatchSnapshot();
  })

})