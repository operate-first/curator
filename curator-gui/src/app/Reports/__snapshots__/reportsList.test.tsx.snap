// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportsList tests should render default DashboardTable component 1`] = `
<div>
  <Component
    aria-label="Compact Table"
    caption="List of reports"
    cells={
      Array [
        "Report Name",
        "Pod CPU Usage in Seconds",
        "Network Usage in Megabits per Second",
        "Memory Usage in Gigabytes",
      ]
    }
    key="dataTable"
    rows={
      Array [
        Object {
          "cells": Array [
            <BodyCell>
              <Link
                to="/projectlist/openshift-monitoring"
              >
                openshift-monitoring
              </Link>
            </BodyCell>,
            21.1,
            12,
            0,
          ],
        },
        Object {
          "cells": Array [
            <BodyCell>
              <Link
                to="/projectlist/hello-world"
              >
                hello-world
              </Link>
            </BodyCell>,
            13.4,
            5,
            0,
          ],
        },
        Object {
          "cells": Array [
            <BodyCell>
              <Link
                to="/projectlist/myproject"
              >
                myproject
              </Link>
            </BodyCell>,
            22.4,
            0,
            0,
          ],
        },
        Object {
          "cells": Array [
            <BodyCell>
              <Link
                to="/projectlist/sample-project"
              >
                sample-project
              </Link>
            </BodyCell>,
            33.5,
            6,
            0,
          ],
        },
        Object {
          "cells": Array [
            <BodyCell>
              <Link
                to="/projectlist/sample-project-1"
              >
                sample-project-1
              </Link>
            </BodyCell>,
            16.1,
            7,
            0,
          ],
        },
        Object {
          "cells": Array [
            <BodyCell>
              <Link
                to="/projectlist/sample-project-2"
              >
                sample-project-2
              </Link>
            </BodyCell>,
            17.4,
            8,
            0,
          ],
        },
        Object {
          "cells": Array [
            <BodyCell>
              <Link
                to="/projectlist/sample-project-3"
              >
                sample-project-3
              </Link>
            </BodyCell>,
            15.5,
            9,
            0,
          ],
        },
      ]
    }
    variant="compact"
  >
    <TableHeader />
    <TableBody />
  </Component>
</div>
`;
